kuma/javascript/src/newsletter.jsx

```diff
@@ -0,0 +1,380 @@
+     * to Google Analytics
+     */
+    const closeNewsletter = () => {
+        setShowNewsletter(false);
+        saveNewsletterVisibleState();
+        ga('send', {
+            hitType: 'event',
+            eventCategory: 'newsletter',
+            eventAction: 'prompt',
+            eventLabel: 'hide'
+        });
+    };
+
+    /**
+     * Shows the privacy checkbox when the email
+     * input field receives focus.
+     */
+    const inputFocusHandler = () => {
+        setShowPrivacyCheckbox(true);
+        ga('send', {
+            hitType: 'event',
+            eventCategory: 'newsletter',
+            eventAction: 'prompt',
+            eventLabel: 'focus'
+        });
+    };
+
+    /**
+     * Attempts to subscribe the user to the `app-dev` newsletter. If an
+     * error occurs when attempting to subscribe the user, it will either:
+     * 1. Present the errors returned to the user
+     * 2. If no reason was provided, it will abandon the XHR request and
+     * send the user directly to the subscription page on www.mozilla.org
+     *
+     * On successful subscription, a thank you message will be shown to the
+     * user, and the visibility state will be stored in `localStorage`
+     * @param {Object} event - The MouseEvent object
+     */
+    const submit = event => {
+        let emailInput = emailRef.current;
+        let newsletterForm = newsletterFormRef.current;
+        let privacyPolicyCheckbox = privacyCheckboxRef.current;
+
+        // if skipXHR
+        if (newsletterForm && newsletterForm.dataset['skipXhr']) {
+            /*
+             * An error occured while attempting to subscribe
+             * the user via Ajax, but no specific error was provided. We
+             * therefore just send the user directly to the Mozorg
+             * newsletter subscription page and log the occurence
+             */
+            ga('send', {
+                hitType: 'event',
+                eventCategory: 'newsletter',
+                eventAction: 'progression',
+                eventLabel: 'error-forward'
+            });
+            return true;
+        }
+
+        event.preventDefault();
+        let ajaxRequest = initAjaxRequest('POST', newsletterSubscribeURL, {
+            'X-Requested-With': 'XMLHttpRequest',
+            'Content-type': 'application/x-www-form-urlencoded'
+        });
+
+        let privacy =
+            privacyPolicyCheckbox && privacyPolicyCheckbox.checked
+                ? '&privacy=true'
+                : '';
+        let documentLocation = encodeURIComponent(document.location.href);
+        let emailInputValue =
+            emailInput && emailInput.value
+                ? encodeURIComponent(emailInput.value)
+                : '';
+        let params = `email=${emailInputValue}&newsletters=${newsletterType}${privacy}&fmt=${newsletterFormat}&source_url=${documentLocation}`;
+
+        ajaxRequest.send(params);
+
+        getAjaxResponse(ajaxRequest).then(response => {
```

Comment:
Can we use `async/await`? The function complexity/nesting seems already a bit high, that'd help a lil